### ISSUE TITLE
(c *DockerClient).PullImage does not pull an image if it already exists

### ISSUE THREAD

**Issue by lucymhdavies** [skybet/cali#3]
As per https://github.com/WheresAlice/staticli/issues/9

Cali default behaviour should be to attempt to pull an image before running it.

(c *DockerClient).StartContainer does attempt a pull:

https://github.com/skybet/cali/blob/master/docker.go#L193-L195
```
	if err := c.PullImage(c.Conf.Image); err != nil {
		return "", fmt.Errorf("Failed to fetch image: %s", err)
	}
```

But the PullImage function will actually do nothing if the image already exists:

https://github.com/skybet/cali/blob/master/docker.go#L358-L361

```
// PullImage - Pull an image locally
func (c *DockerClient) PullImage(image string) error {

	if !c.ImageExists(image) {
```

Personally, I feel that default Cali behaviour here should be to try to pull the image regardless.
There might be circumstances where this is unwanted behaviour, for example when you're in a low bandwidth situation, so seems like it should be configurable. Either by the user, or as a config in the app using the Cali library, or both.

For the record, closed-source PSCLI does the same thing.
But we also avoid using "latest" versions of images in PSCLI, which allows us to make changes to images without worrying about breaking behaviour for people with old versions of PSCLI.

**Comment by lucymhdavies** [comment]
Perhaps "always try to pull the image" is not what I meant.

Perhaps what I mean is:
* Check if use has requested not to pull the image (e.g. with --no-pull-images, and the equivalent option in a cali config file)
* Attempt to pull the image (if you already have the image, moby just returns success)
* If that fails, only then check if the user already has a version of the image

**Comment by WheresAlice** [comment]
Ah, I thought this felt familiar.

So after playing around with this today, here's what I think I expect from Cali:

* Running a command will pull a docker image if necessary or use the existing image if available
* Running a special update command (to be defined) will do the equivalent of `docker pull` which will pull a new version of the image if one is available, doing nothing if we already have the latest version.

Optionally I'd like a way of identifying all of the images referenced in the Cali app, so that I can bulk pull/update. But that might be out of scope for this issue.

**Comment by lucymhdavies** [comment]
> I have an idea for how this would work; I'll get a prototype up and running with existing flags.

This is what I came up with:
https://github.com/lucymhdavies/cali/commit/9fb964a26ef08e18467897a23050879438bb676f



Then, in my app I can override the default value:

```
var (
    // Define this here, then all other files in cmd can add subcommands to it
    cli = cali.NewCli("lucli")
)

func init() {
    cli.SetShort("Lu(cy) CLI")
    cli.SetLong("Named after myself, because I'm vain.\nPronounced 'loosely', because while it's cool, its utility is only loosely based in reality.")

    // Override the default values of some cali-provided flags
    cli.SetFlagDefault("json", true)
}
```

**Comment by lucymhdavies** [comment]
Needs some work of course. I don't like using a `map[string]interface{}` for this, as that means we have to do type assertion when getting the values.

And this is kinda a mess:

```
    c.Flags().BoolVarP(&jsonLogs, "json", "j", c.flagDefaults["json"].(bool), "Log in json format")
    myFlags.BindPFlag("json", c.Flags().Lookup("json"))
    myFlags.SetDefault("json", c.flagDefaults["json"].(bool))
```

Might want to wrap that up into our own Flag type

**Comment by lucymhdavies** [comment]
I like your method of doing this, @WheresAlice 

https://github.com/staticli/staticli/blob/master/cmd/update.go#L18-L32

`staticli update -i`, which pulls any image used by staticli you already have downloaded (i.e. updates them to the latest version with that tag).

**Comment by lucymhdavies** [comment]
I'm going to consider this fixed by #38, and figure out if/how I want to add an "always pull image" config to Cali.
